Merge riak_kv sweeper code to develop

nickelization opened the pull request:
There are still some things left to do before we release this code, but it's solid enough to merge it to develop.

The original sweeper code was written by a contractor, Mikael Lixenstrand. His development work still exist in various repos under branches name 'object_ttl'. This builds off of his work, adding lots of additional test coverage, fixing a bunch of bugs, and generally improving the quality of the code.

This PR depends on three other pull requests being merged first:

- [x] https://github.com/basho/riak_repl/pull/765
- [x] https://github.com/basho/riak_pb/pull/216
- [x] https://github.com/basho/riak_core/pull/882

Commit:
Set state to running on spawn

Commit:
Filter asked participants

The user only needs to care about fail and succ

Commit:
Initial object ttl

* new metadata <<"X-Riak-TTL">>
* add is_x_expired(Obj) to filter expired keys
* riak_kv_object_ttl sweep callbacks

Commit:
Do local put during ttl sweep

Commit:
update_last_modified befor local put

and use ?MD_TTL

Commit:
Add restart sweep

When manual sweep on already started sweep. Stop the old one and
start a new one.

Commit:
Move all sweeper code to riak_kv_sweeper

- add eunit tests for do_sweep

Commit:
Use tombstone_grace_period as switch for reap sweep.

Commit:
Add sweep for index when requested

since index could have been added since last tick.

Commit:
Use same data type (flag) for all config

Commit:
Test schedule_sweep

Commit:
Fix dialyzer

Commit:
decrease throttle lookup

Commit:
Unnecessary stop of scheduling during shutdown

Commit:
Remove persistent participants during normal shutdown

Commit:
Initiate sweeps when the user runs sweeper-status.

Commit:
Init random seed

Commit:
Validate sweep interval

Commit:
Remove unused hashtree_pid function.

Commit:
Use application env instead of file.

Commit:
Make sweep-status safer

Commit:
Make sure riak_kv worker receive all messages during sweeps.

Commit:
Make sweep request use sync_spawn_command

to remove timeouts on riak_kv_vnode_master calls.

Commit:
Remove middleman process

- Was used to release AAE lock through monitoring
- Without it we can get better sync between worker processes and
  riak_kv_sweeper.

Commit:
Update eunit for remove of extra process

Commit:
Remove build_tokens

instead create support for lock types.
One new type of locks: build
Build locks are limited by anti_entropy_build_concurrency and
anti_entropy_concurrency.

Commit:
Ignore responses on local put

Commit:
Add extra throttle

Reuse the sweep throttle limit/sleep for modified objects.

Commit:
Spec update

Commit:
Test lock limits

Commit:
Fix for sweeps with no updated objects

If modified_objects stay at 0 we would throttle for every object.

Commit:
Set sweep concurrency to 1

- Since AAE have a build limit that are default 1 it's better
  to also do sweeps sequential to minimize the risk that we
  need to rerun a sweep because it was not able to get a
  aae build lock.

Commit:
Refactor long lines

Commit:
Use hashtree tombstone to filter diffs

When the sweeper runs asynchronously we need AAE to filter diffs
since all copies of the data will eventually get sweeped.

Commit:
Fix for AAE and leveldb

Commit:
Use clique for sweeper cli.

Commit:
Delete and tobstone AAE for expired TTL objects

This use less disk IO since we delete delete directly.
It also makes AAE compare simpler since it's used by both inter cluster
AAE and AAE fullsync. Inter cluster AAE should not exhange tombstone objects with
tombstone AAE but AAE fullsync should replicate all tombstone objects.

Commit:
Fix dialyzer

Commit:
Update riak-admin sweeper status.

Commit:
Sweeper should wait for riak_kv service

Commit:
Fix repl_consistent_object_filter

Commit:
Get lock before accessing hash tree

Commit:
Add object_ttl deps

Commit:
Add pipe->riak_core

Commit:
Fix spelling.

Commit:
update dialyzer.ignore-warnings

Commit:
Fix crash on late update/finish from sweep.

dict:fetch failed for index that are not longer active on node.

Commit:
Use old obj_size throttle as default.

Commit:
Create binary obj in test.

Commit:
Fix for timeout in estimate

Commit:
Reset total size for obj size throttle

Commit:
Use random sweep

Reduce risk of initial collisions.

Commit:
Remove update in destroy_tree.

destroy_tree/1 used close_trees/1 instead of hashtree:close/1.
The local close_tree/1 also updated the tree before closing it.
During load tests this could sometimes take 2min.
Since we destroy the tree we don't need to update the tree first.

Commit:
Don't time out sweep req

We should wait until done. Same as for fold.

Commit:
Add hash_ops for missing user.

reformat crashed since it didn't have a correct hash_ops.

Commit:
Persist sweep results on shutdown

Commit:
Add sweep_concurrency to cuttlefish

Commit:
Handle all unexpected messages

Commit:
Update ttl sweep to delete

Commit:
Add optional sweep window speed up

nickelization committed:
Merge branch 'develop' into sweeper-develop-merge

Conflicts:
	priv/riak_kv.schema
	rebar.config
	src/riak_kv_entropy_manager.erl
	src/riak_kv_index_hashtree.erl
	src/riak_kv_vnode.erl

nickelization committed:
Fix deadlock that snuck in during sweeper merge

We can't call get_version from within the riak_kv_index_hashtree process
or we will be sending a call to ourselves and we will deadlock.

nickelization committed:
Fix rest_url_encoding_test to work with rebar ct

rebar ct fails because the -module(...) directive rest_url_encoding_test
is guarded by -ifdef(TEST) as TEST is not defined by ct (only by
eunit).

Commit:
Fix use include instead of include_lib

Commit:
Fix putting spaces around '!'

Commit:
Change remove unused ct callbacks

Commit:
Fix revert change accidentally checked in

The convention that test cases match '_test$' is used as a convenience
to discover tests and populate the all() functions. Otherwise one would
have to explicitly maintain the list in the all() function.

Commit:
Change extract functions that create mocks

Commit:
Add testcase where worker process folding over backend crashes

This test shows that a crash in the function doing a fold in this case
the function called back for each data item can leave the sweep in
running state. This is possibly an error but not fixed as yet, just
captured for furthur discussion about a possible solution.

Commit:
Add test for sweeper run_interval functionality

The run_interval field in the #sweep_participant{} record is poorly
named. When set it will cause the sweep_participant to be rescheduled
after run_interval seconds have elapsed after the last sweep.

Note that we are receiving the {ok, successfull_sweep, ...}  message,
sent by the mocked successfull_sweep callback function twice because the
scheduler implementation distinguishes between _never run_ sweeps, which
have the highest priority, and those to be scheduled periodically. It's
the second receive_msg calls that tests scheduling via run_interval.

Note also the second recieve_msg calls using 2500 msecs as the wait
timeout instead of the default. It seems to be the smallest value for
which the test completes successfully.

Commit:
Change put send and receive functions for sweeper callbacks together

Complementary functions to communicate with the process (owned by the
vnode) that the fold is running in.

Commit:
Add common test to test target

Commit:
Add test add sweep participant when scheduler running

The successfull and failed_sweep messages now include the module name to
distinguish which participant completed or failed the sweep.

Commit:
Fix make names consistent

nickelization committed:
Fix crash on get_version called with integer index

This gets called indirectly by sweeper code calling get_lock/2 with an
integer index instead of a pid.

nickelization committed:
Log a message when the hashtree code forces sweep

This can interrupt sweeps in progress, so I want to make sure to flag it
in the logs to clarify any confusing behavior that could ensue.

nickelization committed:
Use `proc_lib` when spawning sweeper utility procs

We should really just never be using plain `spawn`. It hides crashes,
and makes debugging harder.

nickelization committed:
Log if weird stuff happens when estimating keys

Not sure if I like that we just silently default to 0 in these cases.
Maybe there's nothing better we can do, but we should at least log this
scenario.

nickelization committed:
Make send_to_sweep_worker asynchronous

I can't find any reason to wait for a reply, since the reply doesn't
really contain any info, and if the reply times out we'll just keep
going anyway.

It's also possible for messages like `stop_sweep` to get lost if they
hit a worker that has recently finished sweeping, which then causes the
main sweeper process to stall out.

This change should simplify the code, avoid the stalls, and speed things
up slightly to boot.

nickelization committed:
Match the `{ok, Acc}` pattern from fold_objects

I'm not sure if this code path even gets used (Raghav has already
discovered one other obvious bug with it). But this should at least fix
the case clause to match the pattern correctly if we ever return
something other that `{async, Work}` or an error.

Commit:
Fix whitespace

Commit:
Add common test suite for riak_kv_sweeper

Implemented a few simple tests cases using meck to mock external
modules.

Commit:
Add test for riak_kv_sweeper:sweep_request

The test uses a mocked backend fold_objects that returns an #sa{} record
and not {async, Work}. The current implementation did not call
inform_participants to set the sweep result in this case - FIXED.

Commit:
Fix moved riak_kv_sweeper_SUITE to itest directory

By convention eunit tries to auto discover tests in the test directory
and thinks all functions of arity 0 suffixed with _test. Moving the
common tests to a different directory avoids accidentaly discovery. Note
common testcases are functions with arity 1 but there can be an optional
info function with the same of 0 arity that returns properties of the
test cases. The problem was discovered because some testscases were
incorrectly created as 0 arity functions - FIXED.

Commit:
Change all_tests after suggestion from @nm

Using a list comprehension and a helper function instead of a
the hard to read anonymous fold function.

Commit:
Change extract function creating mock riak_core modules

Commit:
Change move sweeper callback functions to their own module

It is confusing when reading the riak_kv_sweeper code to have callback
functions that execute in a riak_kv_vnode owner process mixed together
with those from the riak_kv_sweeper gen_server code. For example, one
can not know which process self() just by looking at the function in
isolation or why we are using gen_server:cast(...) in some of the
function.

The module with the callbacks is named riak_kv_sweeper_fold for lack of
a better name.

Commit:
Add test for stop_all_sweeps

The stop_all_sweeps function also tries to stops all running sweeps by
sending the process that is running the callback function a
message. Messages are handled after a fixed number (1000) keys have been
iterated over. To test stopping a running sweep we need to create a
sweep that iterates over > 1000 keys and runs long enough to receive
stop message and therefore introduce a new visit_function what waits a
given period after iterating over a given number of keys for this test.

Commit:
Fix formatting

Commit:
Add test case to stop all sweeps when the scheduler is running

Implementing this test exposed a bug where the timer event was not
canceled when running stop_all_sweeps.

Commit:
Change rename and reoder lines in test

Commit:
Change use function instead of hardcoded value

The explaining comment is part of the functions documentation

Commit:
Fix correct callback function name

Commit:
Add test case to remove sweep participant when scheduler is running

remove_sweep_participant will try to stop all running sweeps for the
given participant. To test this we need a new visit_function to ensure
we are in a running sweep. The new visit_function synchronizes with a
controller process, visit_function_controller, that synchronizes with
the test case; so the test case receives a message that the sweep on an
index is now waiting and can remove the sweep participant and let the
sweep continue.

The visit_function_controller is required because the scheduler picks an
index to sweep at random. So we can not wait in a fixed order of
indices. An alternate to the visit_function_controller process
implementation is a callback to visit_function where one can remove the
sweep participant.

This test discovered a bug we did not handle throw correctly when
folding over the keys - FIXED.

Commit:
Add logs directory to .gitignore

Commit:
Add test for restart sweep

Making a riak_kv_sweeper:sweep(...) request restarts a running
sweep. Calling sweep(...) multiple does not queue sweeps instead

- If sweep is in the running state, restart it
- If sweep is in the restart state, do nothing

The visit_function_controller changed synchronize multiple times with
visit_function, instead of existing after synchronizing once.

Commit:
Change add sleep before checking result

To account for sweep result message propogation delay.

Commit:
Add test for queue sweep

We need to let the complete on all indices at least once because never
run sweeps are priortized over queued sweeps. Even if a sweep is queued
if it has never been run before it will not be scheduled as a queued
sweep but as never run sweep.

Commit:
Change remove set of sweep_window

Already done in init_per_testcase

Commit:
Add tests for sweep window

Commit:
Fix case clause

It looks like this clause is only used in unit testing.

Commit:
Fix use add_hours function

Commit:
Change do meck:unload(...) in init_per_testcase

Useful when running in an erlang shell using ct:run_test(...) and common
test crashes e.g., meck crashes, and  end_per_testcase(...) is not called.

Commit:
Add tests for get_estimated_keys

Asserts correctness by checking that mocked functions are called since
there is no other way to observe the result of get_estimated_keys.

Commit:
Change used the same prefix as other meck functions

Commit:
Add tests cases for obj_size throttle strategy

Throttling has no observable effect except for waiting, so to test
without resorting to mocking, a variable tracking the total throttle
wait time is added. The obj_size throttling strategy depends on two
parameters

 - amount of data read
 - number mutations to an object during the sweep

When counting amount of data read and overhead of 36 bytes headers in
the serialized riak object is included.

When counting the number of mutations for extra throttling there there
is a hard coded value of 100 mutated objects that triggers an extra
throttle.

The tests found a bug in the with incorrect precedence of + and
rem when counting mutated objects - FIXED.

Commit:
Add tests for pace throttle strategy

Pace throttle strategy waits for a given time after a configurable
number N, of objects are processed. It does an extra throttle on every N
mutated/deleted objects.

Commit:
Fix remove debug print statements

Commit:
Fix use consistent prefix meck_

Commit:
Add test for sweep concurrency > 1

Commit:
Change scheduler to make testable

Refactoring the scheduler to introduce a pure function schedule_sweep2,
and the quickcheck property for it.

The numbered functions schedule_sweep1, schedule_sweep2 is someting I
saw when browsing through common_test sources and personally like
instead of having to invent (meaningless) prefixes/suffixes.

Commit:
Change removed visit_function_controller process

Thanks to a good review comment from @nickelization we can remove the
some complexity from all the test cases

Commit:
Fix setup mock before enabling scheduling

Commit:
Fix revert change incorrectly handling exceptions

Thanks to a review comment from @nickelization we now change to mack
backend and riak_kv_vnode to behave like the riak_kv_worker

Commit:
Fix revert accidentally committed test code

Commit:
Fix setup mocks before enabling the scheduler

Commit:
Fix handle case {ok, Acc} return by Backend:fold_objects(...)

Review comment by @nickelization - the backend objects backend API may
return {ok, Acc} so handling it and added test for this case.

Commit:
Fix match against ok to assert

Commit:
Change remove duplicate testcase

Commit:
Fix spec os:timestamp() not an erlang type

Commit:
Fix test for commit 4444fb8

Commit:
Change make concurrency limit and sweep window part of scheduler

Refactoring to move logic for concurrency limit reached and the sweep
window to the sweep_schedule2 function and extend the quick property to
verify this logic.

Commit:
Fix infinite recursion calling sweep(...) with invalid index

Commit:
Change move logic for sweep request to scheduler

Refactor sweep_request(...). The logic for scheduling a request was
spread on two paths for ticks and requests making it difficult
understand and test and fragile. For example, the concurrency limit
check was on two different paths in the code. Now all logic for
scheduling, either a tick or an explicit request, are in one functions
with EQC properties for it.

Commit:
Change remove maybe_initiate_sweeps from status call

After discussion with @nickelization it seems wrong to have a side
effect of initilization sweeps with the status call. Instead we ensure
that the sweeper is initialized before a sweep.

Commit:
Change removed retries from sweep_request

Commit:
Change initialization and handling of index changes

The sweeper is initialized and we watch for ring changes on
 - on startup in init(...)
 - on handling every tick
 - on handling a sweep_request(...)

Commit:
Change sweeper tests to remove timer:sleep(...)

Working on this commit I expected all tests using the sleeps to run
faster but that is not the case. For repeating sweep_participants with
run_interval set the minimum reaction time for the scheduler is 1
second. We could consider making this configurable to make the tests
faster.

Commit:
Fix dialyzer warnings

There is anomalous dialzyer behaviour

- on line 358, schedule_sweep1({request, ...) dialyzer warns of return
{ok, {tick, ...}} tuples
- line 264 schedule_sweep1(tick, ...) dialyzer does not warn the same

Commit:
Add function for hour generator

Commit:
Change riak_kv_sweeper to use an opaque state

Step 1 for introducing an opaque state type for the sweeper. #sweeps{}
and #sweep_participant{} records still exposed, e.g.,
add_sweep_participant(...) and riak_kv_sweeper_cli.

Using a pattern similar to [webmachine](https://github.com/basho/webmachine/blob/develop/src/webmachine_request.erl)
where internal state is represented as OpaqueState = {?MODULE, State}
allowing calls like:
    ...
    State = riak_kv_sweeper_state:new(),
    State1 = State:update_timer_ref(Ref),
    ...

nickelization committed:
Remove redundant disable_sweep_scheduling API

This is unused, and is doing the exact same thing as stop_all_sweeps.

nickelization committed:
Lots of fixes to whitespace, long lines, etc.

nickelization committed:
Switch to use atoms for fun types, not integers

This is more idiomatic and flexible than a bunch of -defines mapping
to integers.

nickelization committed:
Update tools.mk to the latest version from master

This gives us Common Test support under the test target.

nickelization committed:
Check in sweeper tests for errant messages

I've seen a couple intermittent test failures which I suspect may have
been caused by messages leaking between tests. Haven't been able to
reproduce the issue, but if it happens again this should help confirm or
deny my theory.

nickelization committed:
Remove some unused variables from sweep_request

nickelization committed:
Rename schedule_sweep1 to schedule_sweep

Not sure why we had that 1 on the end there, but it doesn't seem
necessary since there's no schedule_sweep function that would generate a
name conflict.

nickelization committed:
Use an explicit label when nobody has a lock

Using `undefined` is more error-prone and less clear

nickelization committed:
Add locking pid to index hashtree lock state

This isn't used yet, but will be useful in the next commit when we add a
release_lock API function.

nickelization committed:
Add release_lock API call to index hashtree module

Prior to this call being added, the only way to release the lock was for
the locking process to die. This necessitated some ugly hacks in the
sweeper module.

nickelization committed:
Don't use a lock proxy, just release when done

We no longer need to spawn a separate process for locking, since there
now exists an API for manually releasing the lock when we're done with
it.

nickelization committed:
Use lager:warning instead of error_logger

nickelization committed:
In case of invalid sweep window, default to never

If they specified a window, presumably it's because they don't want
sweeps to run during certain times. Since we can't figure out what those
times are, the safe thing is to log a warning and refuse to sweep until
the bad config is fixed.

nickelization committed:
Update sweeper tests for new tree unlock code

nickelization committed:
Always tag the accumulator when throwing

It's confusing and error-prone to throw and catch the accumulator
without any sort of label that we can pattern match on. I also see no
reason to do that when we already have the stop_sweep pattern we can
use.

nickelization committed:
Rename maybe_initiate_sweeps function

There's no "maybe" about it. It will always update the sweep records in
the sweeper state. "Initiate" is also a bad choice of words: we're not
starting any sweeps, as the word initiate implies. We're just updating
the sweeps records in our state which tell us which sweeps we're
responsible for running.

nickelization committed:
Update "clean" build target to remove test beams

nickelization committed:
Tighten up sweeper state module spec

nickelization committed:
Fix dialyzer warnings

nickelization committed:
Fix indentation

nickelization committed:
Correct many incorrect spellings of "successful"

nickelization committed:
Don't stop sweeper ticks when scheduling disabled

This reverts some changes that were introduced in
9fc63e8e08a8b1175e1ce149e0e45b2fe7621683

The ticks do not need to be stopped, since we already refrain from
starting sweeps if sweep scheduling is disabled. Additionally, this
change failed to restart the sweeper ticks if/when scheduling was
re-enabled.

nickelization committed:
Add riak_kv_sweeper behavior specs

nickelization committed:
Rename ask_participants to get_active_participants

The name "ask_participants" seems weird and confusing to me. What are we
asking them? This new name seems much clearer.

nickelization committed:
Remove obsolete wrapper around sweeper state data

We no longer need the {?MODULE, State} construct since we're not using
the State:update_whatever style of coding anymore.

Commit:
Add CT_OPTS variable for make target ct

Example, run a single test case like:
make CT_OPTS='suites=riak_kv_sweeper case=initiailize_sweep_request_test'

Commit:
Use #state{} as first argument to exported functions

As a continuation to 3f84f3f, it is more symmetric to have #state{} as
the first argument instead of the last. Making it the last argument
was required for State:update_whatever pattern.

Commit:
Fix indentation

Commit:
Compile riak_kv_sweeper first

Otherwise -behaviour(riak_kv_sweeper) directives in other modules will
fail compilation if riak_kv_sweeper does not happen to get compiled
first.

Commit:
Don't use maybe_receive_request to wait while throttling

Messages to interrupt a running sweep were received after processing
1000 keys and on every throttle. Receiving interrupt messages on every
throttle is a premature optimization and also hard to test so this
changes removes it.

Commit:
Fix eunit tests after changes in d74f2c7

nickelization committed:
Rewrite sweep to use fold instead of recursion

This is a big refactor of the code that applies the sweep participant
funs. It was previously a big mass of recursive functions and case
statements; this change shifts to using lists:foldl, as well as a bunch
of descriptively-named helper functions.

Commit:
Remove logging of leftover messages, unnecessary, see below

http://erlang.org/documentation/doc-5.6/lib/common_test-1.3.0/doc/html/write_test_chapter.html

Commit:
Throw stop_sweep in one place at the top level fold_req_function

IMO, makes the function and it's spec easier to understand instead of
having two places where we throw stop sweep under the same condition
of no remaining active sweep participants.

Commit:
Add test to check ordering of sweep_participants

nickelization committed:
Remove code for persisting sweeps

Talked with Raghav, and we really can't come up with any good reason why
this was here in the first place. At the very least, we should probably
set an app env or something instead, since if the node goes down there
will clearly be no more sweeps running anyway. However, even then, we
don't have any way of knowing how the state of the system may have
changed between the sweeper process crashing and coming back up, so
blindly restoring the previous state seems highly questionable.

nickelization committed:
Fix regression in bucket props sweeper code

Commit:
Add exometer tracking for keys and bytes swept

Using an exometer sprial probe we report the keys swept and the bytes
swept per index. The spiral will sum the reported values for an
interval giving us a the rate of processing keys and bytes per
index. The effect of sweep concurrency and throttling can also be
visualized by plotting these metrics.

nickelization committed:
Merge branch 'develop' into sweeper-2.3-integration

Conflicts:
	dialyzer.ignore-warnings
	rebar.config
	src/riak_kv_index_hashtree.erl
	src/riak_kv_sup.erl
	src/riak_kv_vnode.erl

nickelization committed:
Fix some previously ignored dialyzer warnings

Not really related to the sweeper branch where we're doing this work,
but Doug and I are feeling caremad ;)

nickelization committed:
Remove some obsolete ignored dialyzer warnings

These aren't happening anymore; no need to suppress them.

Commit:
Add specs for all sweeper functions

Commit:
Fix formatting and long lines

Some code also e.g., renaming variables to wrap long lines to 80 characters

Commit:
Use constant instead of expression

The expresssion 24*3600 should have been in paranthesis to avoid
accidental prcededence bugs. Using a constant also allows us to use
the macro in pattern matches.

Commit:
Use more strict type for module

nickelization committed:
Switch branches back to develop in prep for merge

Commit:
Add more stats for a sweep per index

- num failed sweep participants
- num successful sweep participants
- num mutated objects
- num deleted objects

lehoff commented on src/riak_kv_delete.erl:
```diff
@@ -189,11 +199,61 @@ get_w_options(Bucket, Options) ->
+	Now = os:timestamp(),
+
+	Contents = riak_object:get_contents(RObj),
+	case [{M, V} || {M, V} <- Contents,
```
Running through the entire list should only be done when all the content is in_grace.

Leverage `lists:any/2` by writing something like this:

```erlang
obj_outside_grace_period(RObj, TombstoneGracePeriod) ->
    Now = os:timestamp(),

    Contents = riak_object:get_contents(RObj),
    list:any( fun(Content) -> not content_in_grace(Content, Now, TombstoneGracePeriod) end,
                 Contents).

content_in_grace({Metadata, Value}, Now, TombstoneGracePeriod) ->
    in_grace(Metadata, Now , TombstoneGracePeriod).
```

This way the code reads exactly what the logic is and it will stop as soon as it finds anything outside the grace period.

nickelization replied:
Good catch. Fixed! (Although I used `lists:all` instead of `lists:any` since I believe we only want to reap the object if all the tombstones are outside the grace period, not just one.)

lehoff commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -507,47 +537,64 @@ add_hashtree_pid(true, Index, Pid, State=#state{trees=Trees, trees_version=VTree
+check_lock_type(_Type, _Locks) ->
+    ok.
+
+get_nr_of_locks(Type, Locks) ->
```
Unfortunately, we don't have Elixir's `Enum.count/2` function available, which would fit right in.

Consider not building the list only to measure its length:
```erlang
get_nr_of_locks(Type, Locks) ->
    lists:foldl( fun(#lock{type=T}, Count) when Type==T -> Count+1; 
                            (_, Count) -> Count end,
                     0,
                     Locks ).
```

There ought to be a `lists:count/2` function in OTP, but, alas, no such luck.

nickelization replied:
This feels like a premature optimization to me. This function shouldn't be called that often (or on a critical path), and I do feel that the simple one-line implementation we have now is much clearer and more readable than a fold would be.

If we had lists:count I would gladly use that, though. Perhaps we should write our own implementation of it and add it to a utility module somewhere. Thoughts?

lehoff replied:
Having our own `lists:count/2` would be preferred.

nickelization replied:
`riak_core_util:count/2` now merged - see basho/riak_core#888. Will update sweeper code to make use of it tomorrow.

nickelization replied:
Done.

lehoff commented on src/riak_kv_get_core.erl:
```diff
@@ -195,21 +195,28 @@ final_action(GetCore = #getcore{n = N, merged = Merged0, results = Results,
-                  end,
+
+    ReadRepairs =
+        case ObjState of
```
This case should be its own function, e.g., `object_read_repairs(ObjState, MObj)`.

nickelization replied:
Agreed! Done.

lehoff commented on src/riak_kv_get_core.erl:
```diff
@@ -195,21 +195,28 @@ final_action(GetCore = #getcore{n = N, merged = Merged0, results = Results,
+                        []
+                end
+        end,
     Action = case ReadRepairs of
```
This case should also be its own function.
And the `length([xx || {_Idx, {ok, _RObj}} <- Results])` should be a function that uses the non-existant `lists:count/2` function (we might want to add that function in a util module).
But, regardless, a function called `count_sucessful_results(Results)` should be introduced, no matter how it is implemented.

nickelization replied:
Agreed, and now done using the new `riak_core_util:count/2` func.

lehoff commented on src/riak_kv_get_core.erl:
```diff
@@ -233,6 +240,18 @@ final_action(GetCore = #getcore{n = N, merged = Merged0, results = Results,
+%% dominated (in need of repair) indexes and
+%% the list of not_found (in need of repair)
+%% indexes.
+get_read_repairs(Results, MObj) ->
```
Instead of iterating twice over the `Results` list, consider using `lists:filtermap/2` along these lines:

```erlang
get_read_repairs(Results, MObj) ->
    lists:filtermap( fun(Res) -> index_read_repair(Res, MObj) end, Results ).

index_read_repair({Idx, {ok, RObj}}, MObj) ->
    case riak_object:strict_descendant(MObj, RObj) of
         true ->
            {Idx, outofdate}
        false ->
            false
   end;
index_read_repair({Idx, {error,notfound}}, _MObj) ->
    {Idx, notfound}.
```

nickelization replied:
Good call. Done.

lehoff commented on src/riak_kv_object_ttl.erl:
```diff
@@ -0,0 +1,60 @@
+    InitialAcc = 0,
+    {ok, ttl_fun(), InitialAcc}.
+
+ttl_fun() ->
```
This is a level of unnecessary indirection and it blurs the fact that the function that the `participate_in_sweep/2` has to return has arity 3. 

Change to something like this:
```erlang
participate_in_sweep(_Index, _Pid) ->
    InitialAcc = 0,
    {ok, ttl_fun/3, InitialAcc}.

ttl_fun({{_Bucket, _Key}, RObj}, Acc, [{bucket_props, BucketProps}]) ->
        case riak_kv_util:is_x_expired(RObj, BucketProps) of
            true ->
                 {deleted, Acc};
            _ ->
                {ok, Acc}
        end.
```

Furthermore, this also gives the opportunity to add a spec on the ttl_fun/3, if one so desires.

nickelization replied:
Nice catch. Fixed.

lehoff commented on src/riak_kv_sweeper.erl:
```diff
@@ -0,0 +1,369 @@
+    From ! {ok, Ref},
+    {noreply, State1};
+
+handle_info(sweep_tick, State) ->
```
It is a good idea to remove internal timers as much as possible to make testing easier.
See https://github.com/lehoff/chronos for a deeper explanation.

In this case adding a `trigger_sweep/0` function to the API will remove the need for `handle_info({test_tick...`.

And `schedule_sweep_tick/0` should be changed to use `timer:apply_after/4` or create a module to handle this, alternatively use `chronos`.

nickelization replied:
Just pushed some commits in which I switched the sweeper to use `chronos`, and updated all the CT tests to manually execute sweep ticks instead of waiting for the sweeper to tick itself.

lehoff commented on src/riak_kv_sweeper.erl:
```diff
@@ -0,0 +1,369 @@
+get_estimtate(Index) ->
+    case riak_kv_index_hashtree:get_lock(Index, estimate) of
+        ok ->
+            Result = case riak_kv_index_hashtree:estimate_keys(Index) of
```
Put the case in a function, e.g, `hashtree_estimate_keys(Index)`.

nickelization replied:
Done.

lehoff commented on src/riak_kv_sweeper_cli.erl:
```diff
@@ -0,0 +1,180 @@
+    Header = io_lib:format("~s", ["Sweep results:"]),
+    SortedSweeps = lists:keysort(#sweep.index, Sweeps),
+    Rows =
+        [begin
```
Make the `begin ... end` a separate function.

nickelization replied:
Done (and the code was moved into the riak_kv_sweeper_state module as part of the encapsulation of `#sweep{}`).

lehoff commented on src/riak_kv_sweeper_cli.erl:
```diff
@@ -0,0 +1,180 @@
+                       estimated_keys = {EstimatedKeys, _TS},
+                       swept_keys = SweptKeys},
+                IndexedParticipants) ->
+    case EstimatedKeys of
```
Create a helper function to extract the estimated keys without a case:
```erlang
format_progress(#sweep{index = Index,
                       active_participants = AcitvePart,
                       estimated_keys = {EstimatedKeys, _TS},
                       swept_keys = SweptKeys},
                IndexedParticipants) ->
   EstimatedKeysFormat = format_estimated_keys(EstimatedKeys),
   [{"Index", Index},
    {"Swept keys", SweptKeys}] ++
   EstimatedKeysFormat ++
   format_active_participants(AcitvePart, IndexedParticipants);

format_progress(_, _) ->
    "".

format_estimated_keys(EstimatedKeys) when is_integer(EstimatedKeys)
                                                                             andalso EstimatedKeys > 0 ->
    [{"Estimated Keys", EstimatedKeys}];
format_estimated_keys(_) ->
    [].
```

nickelization replied:
Thanks, did this a while ago, and just noticed I did not reply to the review comment until now. Done.

lehoff commented on src/riak_kv_sweeper_cli.erl:
```diff
@@ -0,0 +1,180 @@
+                     || {Index, Participant} <- IndexedParticipants]).
+
+get_result(Index, Participant, Results, Now) ->
+    case dict:find(Participant#sweep_participant.module, Results) of
```
Do not export variables from the case statement.

```erlang
{LastRun, OutcomeString} =
    case ...
```

nickelization replied:
Missed this comment earlier; just pushed a commit that fixes this.

lehoff commented on include/riak_kv_sweeper.hrl:
```diff
@@ -0,0 +1,65 @@
+%% under the License.
+
+%% Used as configuration for sweep participants.
+-record(sweep_participant,
```
`#sweep_participant{}` should be made into its own module with suitable
  accesssors/setters  or be defined in `riak_kv_sweeper_fold`

nickelization replied:
So a couple weeks ago, I already tried to put it into its own module...but my first attempt really made the code a lot uglier and less readable, because then we can no longer pattern match on record fields in function heads. In order to fix this in a nice clean way, we're going to have to rewrite a pretty substantial amount of code.

If I had written this code, I would not have used shared records all over the place, but I am questioning whether it is worth the effort and trouble to rewrite the sweeper at this point, given that it works and we are under a deadline crunch. I can do it, but it will not be a trivial change.

Also consider that these are not records like e.g. riak_object, where they are passed around all over the system. All these shared records are mostly just used by a few tightly-coupled sweeper modules. They are also not passed across node boundaries via RPC, so there should be no risk of upgrade problems involving mixed clusters with differing record definitions.

nickelization replied:
@lehoff Any thoughts/opinions on the above? Do you think it's worth the time and effort to make the change, or should we leave it as-is?

lehoff replied:
Revisiting its usage today... it should not be in its own module. It belongs in `riak_kv_sweeper_fold`. 

Outside `riak_kv_sweeper_fold` the usage is very limited and can be fixed with a few setters/getters - tests have not been considered.
```
src/riak_kv_delete_sup.erl
70:      #sweep_participant{ description = "Object TTL",
89:      #sweep_participant{ description = "Reap tombstones",

src/riak_kv_entropy_manager.erl
244:      #sweep_participant{

src/riak_kv_sweeper.erl
25:%% schedules sweeps.  Anyone can register a #sweep_participant{} with
34:%% #sweep_participant{}.
123:-spec add_sweep_participant(#sweep_participant{}) -> ok.
139:-spec status() -> {[#sweep_participant{}], [#sweep{}]}.
161:                           [#sweep_participant{}],

src/riak_kv_sweeper_cli.erl
136:     || {Index, #sweep_participant{module = Mod}} <- IndexedParticipants].
162:format_sweep_participant(Index, #sweep_participant{module = Mod,

src/riak_kv_sweeper_state.erl
100:                            Participant:: #sweep_participant{}) -> state().
135:                          {Participants :: [#sweep_participant{}],
153:    {'ok', Participants :: [#sweep_participant{}], state()}.
162:        [Participant#sweep_participant{sweep_fun = Fun, acc = InitialAcc} ||
189:                           [#sweep_participant{}],
344:     {Module, #sweep_participant{run_interval = RunInterval}}
525:        {ok, #sweep_participant{run_interval = RunInterval}} ->
676:    #sweep_participant{module = Module,
744:    #sweep_participant{
```

`riak_kv_delete_sup`: a `riak_kv_sweeper_fold:new_participant/5` function to create the record.

`riak_kv_entropy_manager`: use the function above.
 
`riak_kv_sweeper`: use the `riak_kv_sweeper_fold:participant()` type instead of the record in the type declarations.

`riak_kv_sweeper_cli`:  move the formatting to `riak_kv_sweeper_fold`.

`riak_kv_sweeper_state`: 
* Add a function to get the module out. Not sure what the name should be, maybe `participant_module/1`?
* Add a function to set the fun and initial accumulator. Say, `set_sweep_parameters/3`.
* Add a function to set the run interval.

This is doable. It will make the testing in `riak_kv_sweeper_state` a bit harder, but maybe the generators for the `#sweep_participant{}` should be defined alongside the record definition?

nickelization replied:
Done and pushed. All the sweeper records are now encapsulated and `riak_kv_sweeper.hrl` has been deleted 💥

lehoff commented on include/riak_kv_sweeper.hrl:
```diff
@@ -0,0 +1,65 @@
+         fail_reason
+        }).
+
+-record(sweep,
```
`#sweep{}` should be made local to the `riak_kv_sweeper_state` module as it is mostly
  used there and it should be exported as `riak_kv_sweeper_state.t()`.

nickelization replied:
@lehoff Similar to the case with `#sweep_participant{}`, I wish Mikael had written the code differently, but I'm questioning if it's worth the effort to change it now. Thoughts?

lehoff replied:
For `#sweep{}` the changes are not that big, so I think we should do it now rather than later.

Outside `riak_kv_sweeper_state` `#sweep{}` is only used here (disregarding the test):
```
src/riak_kv_sweeper.erl
139:-spec status() -> {[#sweep_participant{}], [#sweep{}]}.

src/riak_kv_sweeper_cli.erl
100:              || #sweep{state = State} = Sweep <- Sweeps, State == running],
110:format_progress(#sweep{index = Index,

src/riak_kv_sweeper_fold.erl
284:                           #sweep{}) -> 'ok' | 'no_pid'.
285:send_to_sweep_worker(Msg, #sweep{pid = Pid}) when is_pid(Pid) ->
289:send_to_sweep_worker(Msg, #sweep{index = Index}) ->
```

`riak_kv_sweeper` is just a type definition.

`riak_kv_sweeper_cli`: first usage can be fixed by adding a function to extract the state.
2nd usage would require moving the formatting into the `riak_kv_sweeper_state` module, which is not unfair given that it is relating to a type - quite a common pattern in OCaml.

`riak_kv_sweeper_fold`: accessor functions to get the pid and index out.

For this record it makes sense to hide it.

nickelization replied:
Fair enough! I will address this today 👍

nickelization replied:
Fixed. Had to make a bunch of additional changes to the CT tests, but it all works now.

lehoff replied:
Perfect!

lehoff commented on include/riak_kv_sweeper.hrl:
```diff
@@ -0,0 +1,65 @@
+        }).
+
+%% Sweep accumulator
+-record(sa,
```
`#sa{}` should be made local to the `riak_kv_sweeper_fold` module as it is only used there.

nickelization replied:
Done. Had to first rework some test code that was peering directly into `#sa{}`, but the resulting tests are better now anyway.

lehoff commented on src/riak_kv_stat.erl:
```diff
@@ -184,6 +184,12 @@ do_update({vnode_put, Idx, USecs}) ->
     ok = create_or_update([P, ?APP, vnode, puts, time], USecs, histogram),
     do_per_index(puts, Idx, USecs);
+do_update({reap_tombstone, _Index}) ->
+    P = ?PFX,
```
Inline all the `?PFX` - no need to define `P` like that just to save a few letters of typing.

nickelization replied:
@lehoff Technically it's not just saving typing, because ?PFX is not a constant: it expands to a call to `riak_core_stat:prefix()` which does an `app_helper:get_env`. Caching the result of this call in `P` should theoretically offer a tiny performance boost. That said, it's probably negligible, and I'd rather have simpler code, so I went ahead and made your suggested changes anyway. But, just wanted to point that out to you in case it changes your mind at all.

lehoff commented on src/riak_kv_sweeper.erl:
```diff
@@ -0,0 +1,369 @@
+%% ====================================================================
+-spec start_link() -> {ok, pid()}.
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
```
Best to create a `?SERVER` macro (that may call ?MODULE in its definition).
`?MODULE` is for designating callbacks and debug things.

nickelization replied:
Done.

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+    successful_sweep(Succ, Index),
+    failed_sweep(Failed, Index).
+
+-spec successful_sweep([#sweep_participant{}], riak_kv_sweeper:index()) -> [].
```
Change the spec so that the return type is `ok` and use `lists:foreach/2` as the list is not used for anything.

nickelization replied:
Done.

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+    [Module:successful_sweep(Index, FinalAcc) ||
+       #sweep_participant{module = Module, acc = FinalAcc} <- Succ].
+
+-spec failed_sweep([#sweep_participant{}], riak_kv_sweeper:index()) -> [].
```
As above: Change the spec so that the return type is ok and use lists:foreach/2 as the list is not used for anything.

nickelization replied:
And done!

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+       #sweep_participant{module = Module, fail_reason = Reason } <- Failed].
+
+failed_sweep(Failed, Index, Reason) ->
+    [Module:failed_sweep(Index, Reason) ||
```
Use `lists:apply/2`.

nickelization replied:
I assume you mean `lists:foreach/2`. Done.

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+
+%% Throttle depending on swept keys.
+-spec maybe_throttle_sweep(binary(), #sa{}) -> #sa{}.
+maybe_throttle_sweep(_RObjBin, #sa{throttle = {pace, _, _}} = Acc0) ->
```
Move the case into the function clause.

nickelization replied:
✅

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+        throttle_total_wait_msecs = TotalWait} = Acc0,
+    ObjSize = byte_size(RObjBin),
+    TotalObjSize1 = ObjSize + TotalObjSize,
+    case (Limit =/= 0) andalso (TotalObjSize1 > Limit) of
```
Move the `Limit` into the function clause since it will point to the false case faster.

nickelization replied:
Done.

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+
+%% Throttle depending on how many objects the sweep modify.
+-spec maybe_extra_throttle(#sa{}) -> #sa{}.
+maybe_extra_throttle(Acc) ->
```
Extract `TotalWait` in the function clause and create a helper function to calculate `Limit` and `Wait`.

nickelization replied:
@lehoff I'm not sure what you're referring to with the `TotalWait` comment. Clarification?

nickelization replied:
I did go ahead and factor out the code to get `Limit` and `Wait` - that part made sense 👍

lehoff replied:
I meant that instead of pattern matching on the first line of the function move that pattern match into the function head.

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+get_sweep_throttle() ->
+    {Type, Limit, Sleep} =
+        app_helper:get_env(riak_kv, sweep_throttle, ?DEFAULT_SWEEP_THROTTLE),
+    case riak_kv_sweeper:in_sweep_window() of
```
Create an `adjust_sleep(Sleep)` function and simplify this to:

```erlang
get_sweep_throttle() ->
    {Type, Limit, Sleep} =
        app_helper:get_env(riak_kv, sweep_throttle, ?DEFAULT_SWEEP_THROTTLE),
    AdjustedSleep = adjust_sleep(Sleep),
    {Type, Limit, Sleep}.

adjust_sleep(Sleep) ->
    case riak_kv_sweeper:in_sweep_window() of
        true ->
            Div = app_helper:get_env(riak_kv, sweep_window_throttle_div,
                                     ?SWEEP_WINDOW_THROTTLE_DIV),
           Sleep div Div;
        false ->
            Sleep
    end.
```

Maybe make the result of `riak_kv_sweeper:in_sweep_window()` the 2nd argument to `adjust_sleep`.

nickelization replied:
Done.

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+maybe_receive_request(#sa{active_p = Active, failed_p = Fail } = Acc) ->
+    receive
+        stop ->
+            Active1 =
```
Make a function for this. Say, `set_fail_reason_for_participants(Active, sweep_stop)`

nickelization replied:
Done.

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+                 #sweep_participant{} = ActiveSP <- Active],
+            Acc#sa{active_p = [],  failed_p = Active1 ++ Fail};
+        {disable, Module} ->
+            case lists:keytake(Module, #sweep_participant.module, Active) of
```
Make this a function, e.g., `maybe_disable_module`.

nickelization replied:
Done.

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+                           NumMutated :: non_neg_integer(),
+                           NumDeleted :: non_neg_integer(),
+                           BucketProps :: []}) ->
+    {riak_object:riak_object(),
```
Create a local record for this accumulator. It is too complicated to have a 6-ary tuple.

nickelization replied:
Agreed. It started as a smaller tuple when I was first refactoring this code into a fold, and it crept up in size without my really paying attention to it. Switched to a record.

lehoff replied:
Thx.

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+
+-spec get_bucket_props(riak_object:bucket(),
+                       BucketProps :: dict()) ->
+    {BucketProps :: [], BucketPropsDict :: dict()}.
```
Do we have a more telling type for `BucketProps`?  Just a list is rather vague.

nickelization replied:
Did some digging, and found that it is actually `[{atom(), any()}]`. Fixed.

lehoff replied:
Perfect.

lehoff commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -0,0 +1,641 @@
+-spec stat_send(riak_kv_sweeper:index(),
+                {'final'| 'in_progress', #sa{}}) -> #sa{}.
+stat_send(Index, {final, Acc}) ->
+    [ riak_kv_stat:update({sweeper, Index, failed, Module, 1})
```
Use `lists:foreach/2` here - the lists are not used.

nickelization replied:
Done.

lehoff reviewed (changes requested):
Mostly structural issues. Should be easy to fix.

nickelization commented:
Thanks for all the review comments, @lehoff! Will address them today or tomorrow once I finish debugging some problems with overload that I dug up yesterday.

nickelization committed:
Handle overload errors in sweeper and AAE code

nickelization committed:
Improve tombstone.grace_period config schema

Using a data type of "flag" would also allow this to be set to "on",
with undefined results.

Additionally, using the atom 'disabled' instead of 'off' improves the
clarity of the config and the code.

nickelization committed:
Fix code formatting in obj_outside_grace_period

nickelization committed:
Use lists:all to improve grace period code

This allows us to short circuit out early if any of values are still in
the grace period, avoids having to build the intermediate list value,
and makes the code a bit clearer.

nickelization committed:
Factor out logic for getting read repairs

nickelization committed:
Factor out more code from final_action

Better to have a separate utility function than a big sprawling case
statement.

nickelization committed:
Rewrite get_nr_locks to use new count/2 util func

Avoids having to generate that intermediate list representation.

nickelization committed:
Factor out logic to count successful results

Also avoid building an intermediate list by using our new count/2 util
function from riak_core.

nickelization committed:
Fix readrepair code to not traverse the list twice

nickelization committed:
Merge branch 'develop' into sweeper-2.3-integration

Conflicts:
	dialyzer.ignore-warnings
	rebar.config

nickelization committed:
Remove pointless indirection in TTL code

nickelization committed:
Fix bad return value in add_sweep_participant

This was causing a few tests to fail until I noticed it just now.

nickelization committed:
Bump timeout to try and fix tests on builder

Since this works locally, I'm guessing maybe the builders are slow
enough that this timeout is an issue...?

nickelization committed:
Tweak riak_core schema wildcard to fix local issue

We shouldn't need to match on more than one subdirectory, and if we use
** then we can end up timing out and failing the test if someone uses
the common pattern of soft linking .. to deps.

nickelization committed:
Use ?SERVER macro in riak_kv_sweeper

More explicit and clear than just using ?MODULE everywhere.

nickelization committed:
Add workaround for meck bug causing test failures

This change works around a meck bug that was producing weird test
failures without any useful error messages. This ultimately turned out
to be triggered by running eunit tests in the riak_kv_stat module prior
to using meck to mock the riak_kv_stat module in the same eunit run. For
info on this bug, see https://github.com/eproxus/meck/issues/133

For reference, this is what the test output looked like:

Pending:
  undefined
    %% Related process exited with reason: killed

Finished in ? seconds
150 tests, 0 failures, 3 cancelled
ERROR: One or more eunit tests failed.

nickelization committed:
Modify riak_kv_sweeper (and tests) to use chronos

nickelization committed:
Fix dialyzer warnings

nickelization committed:
Lots of test fixes for new sweeper timer API

raghavkarol commented on itest/riak_kv_sweeper_SUITE.erl:
```diff
@@ -433,19 +438,13 @@ scheduler_remove_participant_test(Config) ->
 
-    ok = receive_msg({ok, failed_sweep, sweep_observer_1, WaitIndex}, min_scheduler_response_time_msecs()),
-    #sweep{results = Result} = get_sweep_on_index(WaitIndex),
-    {ok,{_, fail}} = dict:find(sweep_observer_1, Result),
```
@nickelization - Think we shouldn't remove the `{ok, {_, fail}}` match. The intentions is to capture the return and document the return value. It can actually be made stricter:  `{ok,{{_, _, _}, fail}}`.

The matching makes this test case unstable, but I think that can be addressed with doing a `match_retry(...)` instead.

raghavkarol commented on itest/riak_kv_sweeper_SUITE.erl:
```diff
@@ -639,34 +662,17 @@ scheduler_estimated_keys_lock_fail_test(Config) ->
     [false = meck:called(riak_kv_index_hashtree, release_lock, [I]) || I <- Indices],
     ok.
 
-scheduler_sweep_concurrency_test(_Config) ->
```
@nickelization - not sure why you've decided to kill this testcase? Is running concurrent sweeps being tested elsewhere that I'm missing?

nickelization replied:
A lot of the tests now run sweeps concurrently, since we bump up the sweep concurrency and spam a bunch of ticks in a row. I figured explicitly testing it is redundant now that it's being implicitly tested everywhere.

raghavkarol replied:
Agreed that a lot of tests are running sweeps concurrently, but we do not validate that the sweeps are indeed running concurrently. I changed to code to hardcode sweep concurrency of 1 and all the tests still passed, while the same will fail with #1580

nickelization committed:
Switch to using an API call for sleep throttling

This allows us to simplify our tests, by mocking the sweep throttle sleep
calls instead of maintaining an extra variable in the sweep accumulator.
Could also be useful if we want to trace sweep throttles with redbug or
the like.

nickelization committed:
Remove throttle_total_wait_msecs field from #sa

This was only being used for testing, and is no longer needed.

nickelization committed:
Move #sa record out of the sweeper header

This is only used in riak_kv_sweeper_fold - no need to share it with
other modules.

nickelization committed:
Inline instances of ?PFX as per review comments

nickelization committed:
Change stat_send to use lists:foreach

No need to generate a list here.

nickelization committed:
Change successful_sweep to use lists:foreach

Again, no need to build a list here.

nickelization committed:
Replace failed_sweep comprehension with foreach

Another place where we don't need to be building a list.

nickelization committed:
Replace failed_sweep/3 comprehension with foreach

nickelization committed:
Fix type spec

nickelization committed:
Refactor maybe_throttle_sweep

Get rid of some case statements; make better use of function heads.

nickelization committed:
Factor out code to get throttle parameters

nickelization committed:
Refactor get_sweep_throttle to use a helper func

nickelization committed:
Refactor maybe_receive_request into smaller funcs

nickelization committed:
Switch sweep fold to use a record accumulator

Much more manageable than a 6-tuple.

raghavkarol commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -200,34 +200,31 @@ format_result(SuccFail, Results) ->
-    end;
-
+maybe_throttle_sweep(_RObjBin, #sa{throttle = {pace, Limit, Wait}, swept_keys = SweepKeys} = Acc)
+  when SweepKeys rem Limit =:= 0 ->
```
The comment 
``` 
%% We use recieve ...
%% This way ...
``` 
Is no longer correct  and should be deleted

nickelization replied:
Ah nice catch, thanks! Fixed.

raghavkarol commented on src/riak_kv_sweeper_fold.erl:
```diff
@@ -465,10 +465,14 @@ update_stat_counters(NumMutated, NumDeleted, RObjBinSize, Acc) ->
 -spec stat_send(riak_kv_sweeper:index(),
                 {'final'| 'in_progress', #sa{}}) -> #sa{}.
 stat_send(Index, {final, Acc}) ->
```
Just a note, list comprehensions [do not generate a list](http://erlang.org/doc/efficiency_guide/listHandling.html#id68285), when the compiler can detect that the return value is not used.

nickelization replied:
Ooh wow! I didn't know that, that's super cool.

@lehoff I already made lots of list comprehension to `lists:foreach` changes at your suggestion, to avoid generating an intermediate list. I'm now wondering if that is a good idea, since Raghav has helpfully pointed out that basically this optimizes down to the same thing in the end. Any objections if I revert some of my commits in cases where a list comprehension is more readable/succinct?

Commit:
Add test for concurrent sweeps

Reintroduce and fix the test removed in 054b62c. This testcase
validates sweep concurrency functionality asserting that an expected
number of sweeps are running in concurrently.

Commit:
Remove unused function

nickelization committed:
Remove obsolete comment in sweeper code

nickelization committed:
Improve type spec on bucket props in sweeper code

nickelization committed:
Fix misspelling of "estimate"

Commit:
Use 1 as first index when creating partitions

Were creating Partitions+1 indices

nickelization committed:
Factor out code to get key estimate from hashtree

Commit:
Update sweep concurrency test

Add checks that we do not start more than sweep_concurrency_limit
sweeps in parallel.

nickelization committed:
Improve type specs on dicts in sweeper state

nickelization committed:
Update several sweeper copyright headers to 2016

nickelization committed:
Make the #sweep record internal to sweeper state

Instead of having a shared public record, this changeset encapsulates
the #sweep record into the riak_kv_sweeper_state module.

Still TODO - fix the CT test suite to work with these changes.

nickelization committed:
Remove unused parse transform for Pulse

No pulse tests here, not sure why we were instrumenting for it.

nickelization committed:
Undo nesting of -ifdefs in sweeper state module

No reason that EQC would be defined when TEST isn't, and it's much
easier to read if the ifdefs aren't nested together like that.

nickelization committed:
Update sweeper tests for encapsulated #sweep{} rec

nickelization committed:
Fix some dialyzer issues

I can't figure out exactly what the issue is with opaque types, but it
doesn't seem super important that we use them anyway.

nickelization committed:
Refactor formatting code for estimated keys

Commit:
Update test to remove usage of #sweep record

Sweep record is now incapsulated in the state, see dd7c4b5.

lehoff commented on src/riak_kv_sweeper.erl:
```diff
@@ -0,0 +1,369 @@
+            {ok, [], State1} ->
+                State1;
+            {ok, ActiveParticipants, State1} ->
+                ?MODULE:update_started_sweep(Index,
```
Doing a `?MODULE:some_fun` call is dubious. It triggers code upgrade, which does not seem to be the wish here.

I think the `?MODULE:` should be dropped.

nickelization replied:
Yeah, no idea why that was there. Dropped.

lehoff commented on src/riak_kv_sweeper_state.erl:
```diff
@@ -0,0 +1,1076 @@
+expired_or_missing(Now, #sweep{results = Results}, Participants) ->
+    ResultsList = dict:to_list(Results),
+    Missing = missing(run_interval, Participants, ResultsList),
+    Expired =
```
Why is this not in a function like `missing/3`?

nickelization replied:
Not sure why not, but it's in its own function now! Fixed.

lehoff commented on src/riak_kv_sweeper_state.erl:
```diff
@@ -0,0 +1,1076 @@
+expired(_Now, _TS, disabled) ->
+    0;
+expired(Now, TS, RunInterval) ->
+    case elapsed_secs(Now, TS) - RunInterval of
```
Replace the case statement with:
`max(0, elapesed_secs(Now, TS) - RunInterval)`

nickelization replied:
Good call. Done!

lehoff commented on src/riak_kv_sweeper_state.erl:
```diff
@@ -0,0 +1,1076 @@
+-spec scheduler_enabled() -> boolean().
+scheduler_enabled() ->
+    case app_helper:get_env(riak_kv, sweeper_scheduler, true) of
+        E when E == true; E == false ->
```
Change the guard to `when is_boolean(E)`

nickelization replied:
And done.

lehoff commented on src/riak_kv_sweeper_state.erl:
```diff
@@ -0,0 +1,1076 @@
+            always;
+        {ok, never} ->
+            never;
+        {ok, {StartHour, EndHour}} when StartHour >= 0, StartHour =< 23,
```
Make a macro:
```erlang
-define(IS_HOUR(H), is_integer(H) andalso H >=0 andalso H <=0).

...
        {ok, {StartHour, EndHour}} when ?IS_HOUR(StartHour), ?IS_HOUR(EndHour) -> 
        ...
```
And perhaps `StartHour =< EndHour`??

nickelization replied:
Macro created. However, it is valid to have `StartHour` > `EndHour`. This allows you to have a sweep window from, say, 10PM to 2AM. The existing sweeper code handles this correctly.

lehoff commented on src/riak_kv_sweeper_state.erl:
```diff
@@ -0,0 +1,1076 @@
+
+-spec get_concurrency_limit() -> non_neg_integer().
+get_concurrency_limit() ->
+    case app_helper:get_env(riak_kv, sweep_concurrency,
```
Are we okay with this failing if `sweep_concurrency` is not an integer in the environment?

nickelization replied:
Normally Cuttlefish should guard against non-integer values here, so it's not a huge risk, but it is possible a user might go screwing around with advanced.config and put a bad value in there. Better safe than sorry, I suppose.

Now that I look at this function, I'm also not sure why we have that `case` statement there at all, since `?DEFAULT_SWEEP_CONCURRENCY` is already an integer. Will fix that as well.

nickelization committed:
Remove unnecessary ?MODULE qualifier in sweeper

nickelization committed:
Factor out code for getting expired counts

nickelization committed:
Replace case statement with call to max in sweeper

nickelization committed:
Use is_boolean instead of matching true/false

nickelization committed:
Use ?IS_HOUR macro to clean up sweeper state code

nickelization committed:
Fix get_concurrency_limit to only return integers

Unlikely that we'd have a bad value here, since we should have
cuttlefish protecting us, but it is totally possible for a user to go
messing around in advanced.config and shooting themselves in the foot.

Better to have a friendlier error to give in such a scenario.

nickelization committed:
Bump sweep fold test timeouts up to 30 seconds

Eunit defaults to a 5 second timeout per test, and these tests were
occasionally exceeding that on the build machine (typical runtime for
some of these tests is around 4 seconds).

nickelization committed:
Switch riak_pb dep branch back for more testing

Trying to get a build set up for performance testing, and since this PR
isn't ready to merge yet I'm switching this back so we can get a working
riak_ee branch going.

Commit:
Extract repeated lines into function

nickelization committed:
Add .ackrc to .gitignore

Having a local .ackrc is particularly useful now, as a way to ensure we
ignore the contents of the Common Test log directory.

nickelization commented:
thumbot retry

nickelization committed:
Encapsulate #sweep_participant in sweeper fold mod

This kills another shared record, and allows us to entirely remove the
riak_kv_sweeper.hrl file.

Still need to update the CT suite to work with these changes.

nickelization committed:
Use not_found instead of [] for no expired sweeps

Returning an empty list to indicate no expired sweeps were found seems
slightly insane when the function does not return a list under any other
scenarios....

nickelization committed:
Fix sweeper CT suite for encapsulated participant

Now that the #sweep_participant record is no longer public, these
changes are needed to fix the test suite.

nickelization committed:
Fix 31 misspellings of the word "participant"

Ah, the dangers of autocompletion ;)

nickelization committed:
Rename meck_new_sweep_participant test util func

It's no longer just making meck calls, so the new name avoids any
potential confusion. We can also factor out the meck calls into their
own utility function.

nickelization committed:
Fix sweep concurrency test for latest rec changes

Now that sweep_participant is encapsulated in riak_kv_sweeper_fold, this
test case needs to be updated to match.

nickelization committed:
Simplify sweeper get_active_participants function

No need to return the state since we never modify it. Also no need to
wrap things in an {ok, ...} tuple since we never return an error.

nickelization committed:
Add sweep acc as arg to failed_sweep callback

This may be useful sometimes; no reason not to provide access to it.
Thanks to @angrycub for the suggestion.

nickelization committed:
Merge pull request #1580 from basho/rkarol-sweep_concurrency_test

Add test for concurrent sweeps

nickelization committed:
Fix sweeper CT suite for new failed_sweep API

nickelization committed:
Don't export vars from case in sweeper CLI code

nickelization commented:
thumbot retry

nickelization committed:
Fix sweeper fold test for new failed_sweep API

Also improving variable naming while I'm in there.

davidx commented:
thumbot retry

nickelization commented:
Discussed with @lehoff this morning, and we've decided there's been enough review and refinement of this work, and it's time to merge it. The dialyzer step is still failing for Thumbs, but that's due to an issue in the develop branch.
